### ISSUE TITLE
Identify use cases to propagate and adjust Element Arrangement.

### ISSUE THREAD

**Issue by moriohara** [torch-spyre/torch-spyre#4080]
Data type conversion ops to change the element byte size, such as fp16 to fp32 and vice versa, affect the order of elements at the stick dimension, aka Element Arrangement or EA. A torch-level operation can take one or more input tensors when their EA meets a certain condition. Otherwise, the EA needs to be adjusted very likely with an extra memory access overhead. To develop a compilation algorithm to minimize such overhead, we want to understand how the EA can propagate across ops and when the EA needs to be adjusted.
This issue is an extension to other existing issues, such as https://github.com/torch-spyre/torch-spyre/issues/3730, https://github.com/torch-spyre/torch-spyre/issues/2508 and [RFC 2971](https://github.com/torch-spyre/RFCs/blob/main/2971-FP32ElementArrangement/2971-FP32ElementArrangementRFC.md)

**Comment by moriohara** [comment]
## Initial Analysis

Three district patterns, normalization with upcast/downcast, comparison to mask, and index to experts, have been identified from the nine target models: 
[Llama-3.1-8B](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/Meta-Llama-3.1-8B-Instruct.yaml), [Ministral-3-14B](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/Ministral-3-14B-Instruct-2512.yaml), [Mistral-Small-3.2-24B](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/Mistral-Small-3.2-24B-Instruct-2506.yaml), [Qwen2.5-7B](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/Qwen2.5-7B-Instruct.yaml), [gemma-4-26B](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/gemma-4-26B-A4B-it.yaml), [gpt-oss-20b](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/gpt-oss-20b.yaml), [granite-3.3-8b-fms](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/granite-3.3-8b-instruct-fms.yaml), [granite-3.3-8b](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/granite-3.3-8b-instruct.yaml), [granite-4.1-8b](https://github.com/torch-spyre/torch-spyre/blob/main/tests/resource/models/granite-4.1-8b.yaml)
 
### Normalization with upcast/downcast 
Operations, like RMSNorm, are typically associated with precision upcast/downcast to avoid precision errors due to a reduction and other arithmetic ops. While upcast/downcast affects the EA, this class of use cases typically don't require an EA adjustment because fp32 tensors are ephemeral, a reduction op is agnostic to the EA (the inner function is considered as commutative), and a point-wise op with broadcast is also agnostic to the EA. More details can be found in [RFC 2971](https://github.com/torch-spyre/RFCs/blob/main/2971-FP32ElementArrangement/2971-FP32ElementArrangementRFC.md) and https://github.com/torch-spyre/torch-spyre/issues/2508.

### Comparison to mask
Operations, like torch.le/ge/eq, produces a bool tensor. A equivalent native op on Spyre takes either fp16 or fp32 tensors to produces a bool tensor with the same dtype as the inputs. The output is consumed at a mask op as a mask (e.g. `masked_fill_`, `masked_scatter`, `where`). A Spyre-native mask op requires all tensor inputs have the same dtype (fp16 or fp32), which can be different from that of inputs of the comparison op. If different, the dtype needs to be converted, hence the EA needs to be adjusted. A custom op to restore the EA of a staggered fp16 tensor is already available (https://github.com/torch-spyre/torch-spyre/pull/3571). More details about these use cases can be found in https://github.com/torch-spyre/torch-spyre/issues/3730#issuecomment-5432639184.

### Index to experts (MOE)
An MOE use case involves a gather op to access a tensor using an index of top-k experts. While the gather op takes an int32 index tensor, the top-k op may involve fp16 or fp32 arithmetic operations internally. Here, out of the nine models, `gemma` and `gpt-oss` provide a similar but distinctive use case:

#### gpt-oss
The topk op in gpt-oss takes an fp16 tensor to obtain the top-k experts. It internally relies on fp16 arithmetic ops to produce a pair of fp16 value and index tensors. The latter is to be converted to fp32 as an internal representation of integer values. This conversion also affects the EA to make the index tensor staggered. A torch.getitem op takes this index tensor to perform a gather op. The EA needs to be restored to obtain the correct result. A custom op similar to https://github.com/torch-spyre/torch-spyre/pull/3571 can be applied to the fp16 index tensor from the internal topk op in such a way that the fp32 tensor converted from fp16 has the standard EA.

https://github.com/huggingface/transformers/blob/5eddc12edfaf8cafde8c9bae4ccb12f8a139b4f9/src/transformers/models/gpt_oss/modeling_gpt_oss.py#L128
```python
router_top_value, router_indices = torch.topk(router_logits, self.top_k, dim=-1)  # topk returns indices as fp16 to be converted to fp32/int32
```
https://github.com/huggingface/transformers/blob/5eddc12edfaf8cafde8c9bae4ccb12f8a139b4f9/src/transformers/integrations/moe.py#L126-L139
```python
expert_ids = top_k_index.reshape(-1)  # (S,) # top_k_index ← router_indices ← topk
expert_ids = expert_ids.clamp(0, self.num_experts - 1) # point-wise
selected_weights = self.gate_up_proj[expert_ids] # gather/getitem with router_indices as an index tensor
```

#### gemma
The topk op in gemma takes an fp32 tensor from a softmax with upcast, where the input of the topk is already staggered as the softmax doesn't downcast unlike RMSNorm. As the index tensor from the topk op is computed against the staggered tensor, the content needs to be adjusted before further consumption at other ops. Note that the adjustment maps a tensor index from the staggered tensor to the original tensor through arithmetic operations, less costly than a general EA adjustment which moves elements with a stick.

https://github.com/huggingface/transformers/blob/a08ace4bbd97e721c98751deec37d87b026acadc/src/transformers/models/gemma4/modeling_gemma4.py#L1358-L1371
``` python
router_probabilities = nn.functional.softmax(expert_scores, dim=-1, dtype=torch.float32) # expert_scores=[34, 128] fp16 to be upcasted (staggered)

# topk returns both values (probabilities) and indices directly
top_k_weights, top_k_index = torch.topk(
            router_probabilities,
            k=self.config.top_k_experts,
            dim=-1,
 )  # both [B*S, K] returning fp32 weights and int64 index (is index STANDARD EA?)

top_k_weights = top_k_weights * self.per_expert_scale[top_k_index] # gather/getitem with top_k_index
```
https://github.com/huggingface/transformers/blob/a08ace4bbd97e721c98751deec37d87b026acadc/src/transformers/integrations/moe.py#L396-L399
```python
expert_ids = top_k_index.reshape(-1)  # (S,)
# Sort by expert for grouped processing
expert_ids_g, perm = torch.sort(expert_ids)
```
https://github.com/huggingface/transformers/blob/a08ace4bbd97e721c98751deec37d87b026acadc/src/transformers/integrations/moe.py#L407-L409
```python
histc_input = expert_ids_g.float() if device.type in ("cpu", "mps") else expert_ids_g.int()
tokens_per_expert = torch.histc(histc_input, bins=self.num_experts, min=0, max=self.num_experts - 1)
offsets = torch.cumsum(tokens_per_expert, dim=0, dtype=torch.int32) # fp32 cumsum
```
https://github.com/huggingface/transformers/blob/a08ace4bbd97e721c98751deec37d87b026acadc/src/transformers/integrations/moe.py#L426-L427
```python
sentinel_mask = (expert_ids_g >= self.num_experts).unsqueeze(-1)
expert_ids_g.clamp_(max=self.num_experts - 1)
```
